src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -108,6 +122,94 @@ ShouldCreateTenantSchemaTable(Oid relationId)
 }
 
 
+/*
+ * EnsureTableKindSupportedForTenantSchema ensures that given table's kind is
+ * supported by a tenant schema.
+ */
+static void
+EnsureTableKindSupportedForTenantSchema(Oid relationId)
+{
+	if (IsForeignTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot create a foreign table in a distributed "
+							   "schema")));
+	}
+	else if (PartitionTable(relationId))
+	{
+		ErrorIfIllegalPartitioningInTenantSchema(PartitionParentOid(relationId),
+												 relationId);
+	}
+	else if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errmsg("tables in a distributed schema cannot inherit or "
+							   "be inherited")));
+	}
+}
+
+
+/*
+ * EnsureFKeysForTenantTable ensures that all referencing and referenced foreign
+ * keys are allowed for given table.
+ */
+static void
+EnsureFKeysForTenantTable(Oid relationId)
+{
+	Oid tenantSchemaId = get_rel_namespace(relationId);
+	int fKeyReferencingFlags = INCLUDE_REFERENCING_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
+	List *referencingForeignKeys = GetForeignKeyOids(relationId, fKeyReferencingFlags);
+	Oid foreignKeyId = InvalidOid;
+	foreach_oid(foreignKeyId, referencingForeignKeys)
+	{
+		Oid referencedTableId = GetReferencedTableId(foreignKeyId);
+		Oid referencedTableSchemaId = get_rel_namespace(referencedTableId);
+
+		/* We allow foreign keys to a table in the same schema */
+		if (tenantSchemaId == referencedTableSchemaId)
+		{
+			continue;
+		}
+
+		/*
+		 * Allow foreign keys to the other schema only if the referenced table is
+		 * a reference table.
+		 */
+		if (!IsCitusTable(referencedTableId) ||
+			!IsCitusTableType(referencedTableId, REFERENCE_TABLE))
+		{
+			ereport(ERROR, errmsg("when referenced table is in another schema, "
+								  "only foreign keys to a reference table is "
+								  "allowed from a table in distributed schema"));
+		}
+	}
+
+	int fKeyReferencedFlags = INCLUDE_REFERENCED_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
+	List *referencedForeignKeys = GetForeignKeyOids(relationId, fKeyReferencedFlags);
+	foreach_oid(foreignKeyId, referencedForeignKeys)
+	{
+		Oid referencingTableId = GetReferencingTableId(foreignKeyId);
+		Oid referencingTableSchemaId = get_rel_namespace(referencingTableId);
+
+		/* We allow foreign keys from a table in the same schema */
+		if (tenantSchemaId == referencingTableSchemaId)
+		{
+			continue;
+		}
+
+		/*
+		 * Allow foreign keys from the other schema only if the referencing table is
+		 * a reference table.
+		 */
+		if (!IsCitusTable(referencingTableId) ||
+			!IsCitusTableType(referencingTableId, REFERENCE_TABLE))
+		{
+			ereport(ERROR, errmsg("when referencing table is in another schema, "
+								  "only foreign keys from a reference table is "
+								  "allowed to a table in distributed schema"));
+		}
```

Comment:
A reference table cannot reference a tenant table (but the other way around is okay --as in above loop), so we should throw an error here without checking the table type.

Also, does that look good to you @aykut-bozkurt /  @mulander?

```suggestion
			ereport(ERROR, errmsg("foreign keys to tables in distributed "
			                      "schemas are not allowed when the referencing "
			                      "table is in another schema"),
			               errdetail("\"%s\" references \"%s\" via foreign key "
			                         "constraint" \"%s\",
			                         generate_qualified_relation_name, 
			                         generate_qualified_relation_name,
			                         get_constraint_name));
```